helix-core/src/main/java/org/apache/helix/controller/stages/MessageDispatchStage.java

```diff
@@ -78,7 +78,17 @@ protected void processEvent(ClusterEvent event, MessageOutput messageOutput) thr
         batchMessage(dataAccessor.keyBuilder(), messagesToSend, resourceMap, liveInstanceMap,
             manager.getProperties());
 
+    String expectedSession = event.getAttribute(AttributeName.EVENT_SESSION.name());
+    // An early check for expected leader session. If the sessions don't match, it means the
+    // controller lost leadership, then messages should not be sent and pipeline should stop.
+    // This avoids potential double masters for a single partition.
+    if (expectedSession != null && !expectedSession.equals(manager.getSessionId())) {
+      throw new StageException(
+          "Controller: " + manager.getInstanceName() + " lost leadership! Expected session: "
+              + expectedSession + ", actual: " + manager.getSessionId());
+    }
```

Comment:
should we do this for all Stage? or just messageDispatchStage?  If for all stages, maybe we should put this check before where each stage was called.

Reply:
We've discussed this question. We decided not to do the check for all stages (too many stages). Performance is not concern. Correctness is the most critical. We only do it for to ensure correctness here.

Reply:
Why checking in every stage is too much? Do you mean code wise is too much, actually you just need the same code, in different place, right?

Reply:
What I was thinking about is, we also check this stage in `ReadClusterDataStage()` because this place is the highly possible one that session expires. In our issue, the session expiry also happened in `ReadClusterDataStage`.

If we don't check it at every stage, it just costs extra resource(CPU/memory/network IO) which we could tolerate. I think checking at every stage may be excessive because most of stages are pretty fast(`ReadClusterDataStage` takes longer time) and very tiny possibility that session changes. As long as we protect sending message, correctness is ensured. We wanted to make minimum code change.